File: frontend/src/types/simulation.ts
// ─── Simulation ──────────────────────────────────────────────────────────────

export interface SimulationModification {
  type: 'duration_scale' | 'remove_activity' | 'adjust_frequency';
  activity: string;
  value: number;
}

export interface SimulationActivityStats {
  name: string;
  frequency: number;
  avg_duration: number;
}

export interface SimulationStats {
  total_cases: number;
  total_events: number;
  avg_case_duration: number;
  median_case_duration: number;
  avg_events_per_case: number;
  activities: SimulationActivityStats[];
}

export interface SimulationResponse {
  original: SimulationStats;
  simulated: SimulationStats;
  improvement: {
    avg_duration_change_pct: number;
    case_count_change: number;
    activities_removed: string[];
  };
}

// ─── Discrete-Event Simulation (DES) ─────────────────────────────────────────

export interface DESActivityDuration {
  mean: number;
  std: number;
  samples: number[];
  count: number;
}

export interface DESResourcePool {
  capacity: number;
  cases_handled: number;
}

export interface DESParameters {
  arrival_distribution: {
    kind: 'exponential' | 'empirical';
    lambda: number;
    mean_inter_arrival_s: number;
  };
  activity_durations: Record<string, DESActivityDuration>;
  gateway_probabilities: Record<string, Record<string, number>>;
  resource_pools: Record<string, DESResourcePool>;
  hourly_calendar: Record<string, number>;
  act_resource_map: Record<string, string | null>;
  start_activities: string[];
  sink_activities: string[];
  total_cases_observed: number;
}

export interface DESScenario {
  arrival_rate_multiplier?: number;
  activity_duration_overrides?: Record<string, number>;
  activity_automation?: Record<string, boolean>;
  resource_pool_overrides?: Record<string, number>;
  new_resources?: Array<{ name: string; capacity: number }>;
}

export interface DESSummary {
  avg_case_duration_s: number;
  p50: number;
  p90: number;
  p95: number;
  throughput_cases_per_day: number;
  max_concurrent_cases: number;
  resource_utilization: Record<string, number>;
}

export interface DESSimulationResult {
  summary: DESSummary;
  baseline: DESSummary;
  delta: Record<string, number>;
  runs: number;
}

File: backend/app/schemas/simulation.py
"""
Pydantic schemas for process simulation / what-if and discrete-event-simulation endpoints.
"""

from uuid import UUID

from pydantic import BaseModel, Field


# --- Process Simulation / What-If Analysis ---


class SimulationModification(BaseModel):
    type: str = Field(
        ...,
        description="Modification type: duration_scale, remove_activity, adjust_frequency",
    )
    activity: str = Field(..., description="Target activity name")
    value: float = Field(
        ...,
        description=(
            "Modification value: scale factor for duration_scale, "
            "percentage (0-100) for adjust_frequency, ignored for remove_activity"
        ),
    )


class SimulationRequest(BaseModel):
    event_log_id: UUID = Field(..., description="Event log to simulate from")
    num_traces: int = Field(
        default=500,
        ge=1,
        le=5000,
        description="Number of synthetic traces to generate via Petri net playout",
    )
    modifications: list[SimulationModification] = Field(
        default=[],
        description="List of what-if modifications to apply to the simulated log",
    )


class SimulationStats(BaseModel):
    total_cases: int = Field(..., description="Number of cases")
    total_events: int = Field(..., description="Total number of events")
    avg_case_duration: float = Field(
        ..., description="Average case duration in seconds"
    )
    median_case_duration: float = Field(
        ..., description="Median case duration in seconds"
    )
    avg_events_per_case: float = Field(..., description="Average events per case")
    activities: list[dict] = Field(
        default=[],
        description="Per-activity stats: [{name, frequency, avg_duration}]",
    )


class SimulationResponse(BaseModel):
    original: SimulationStats
    simulated: SimulationStats
    improvement: dict = Field(
        ...,
        description=(
            "Comparison metrics: {avg_duration_change_pct, "
            "case_count_change, activities_removed}"
        ),
    )


# --- Discrete-Event Simulation (DES) ---


class DESScenario(BaseModel):
    """What-if scenario for the DES engine."""
    arrival_rate_multiplier: float = Field(
        default=1.0,
        description="Multiply arrival rate (>1 = more cases per day)",
    )
    activity_duration_overrides: dict[str, float] = Field(
        default={},
        description="Per-activity duration multiplier (0.5 = 2x faster)",
    )
    activity_automation: dict[str, bool] = Field(
        default={},
        description="Mark activity as automated (duration → 0)",
    )
    resource_pool_overrides: dict[str, int] = Field(
        default={},
        description="Override capacity for named resource pools",
    )
    new_resources: list[dict] = Field(
        default=[],
        description="Add new resource pools [{name: str, capacity: int}]",
    )


class DESSummary(BaseModel):
    avg_case_duration_s: float
    p50: float
    p90: float
    p95: float
    throughput_cases_per_day: float
    max_concurrent_cases: int
    resource_utilization: dict[str, float]


class DESSimulationResponse(BaseModel):
    summary: DESSummary
    baseline: DESSummary
    delta: dict[str, float]
    runs: int
